frame/assets/src/lib.rs

```diff
@@ -193,6 +196,10 @@ pub mod pallet {
 			+ MaxEncodedLen
 			+ TypeInfo;
 
+		/// Max number of storage keys to destroy per extrinsic call.
+		#[pallet::constant]
+		type RemoveKeysLimit: Get<u32>;
```

Comment:
I would prefer the witness data instead.

Because current way it's more complex that might be seen.
It's not really # of storage keys to destroy, but # of some unit of work to be done.
For example destroy_approvals has T::Currency::unreserve within that unit of work, destroy_accounts does not have it. Currency can be anything. 

Having only one limit you assume that these units of work will be relatively similar for different extrinsics. 
If they differ a lot, the clients of this pallet will have to calculate and use the limit for the heaviest one. 

You can have different limits for different extrinsics, but it will have a little difference with having them as an extrinsic argument (witness data).
But the API to the pallet in my opinion a bit more complex (with more limits as config).
And users of extrinsics will have less flexibility.

Reply:
Won't calculating such a witness be complicated? The caller of the extrinsic would need to get the total number of the asset and then needs to know what the max is, so they can call only with less than that value, or call the extrinsic with a very large witness, and it's not clear how we should behave if we get a large witness. How would we even know if the witness is too large from within the extrinsic logic?

Reply:
Yeah, I think it's hard to get around some configured max here, because the witness could be larger than what the system can delete. So even if it only destroys _up to witness_ number, it needs to know if the witness is too big.

I actually kind of like the idea in https://github.com/paritytech/substrate/pull/12310#pullrequestreview-1125106096 to put this in `on_idle` such that the destruction will be carried out by the system later. That could come as a follow up PR.

Reply:
@tonyalaribe what do you think about witness arg instead RemoveKeysLimit for destroy accounts and destroy approvals? 
and what about on_idle?

Reply:
@muharem about using a witness arg, could you explain what you're suggesting? Especially how we will deal with witness being too big and not fitting into a block? I think that's what i'm not clear about. 

About being able to call the functions from on_idle and on_initialize, I added `do_*` versions of the extrinsics that can be called from on_idle and on_initialize (), the `do_*` functions don't require an origin argument, so they can be called on_idle and on_initialize. 
eg: https://github.com/paritytech/substrate/pull/12310/files#diff-b19c80e2768586b8936d8bc731b2cfa5d165dad248bf381be5ae9b52746f9f2dR673

Reply:
@tonyalaribe if witness is too, the weight calc func, will calc the weight which wont fit to the block, and the call will be rejected. 

yes, you can use them within `on_idle`.

Reply:
discussed offline, 
we keep limit set as config for better user experience.
and we use on_idle for lazy destroy after destroy has started